File: rss.pl
#!/usr/bin/perl -w
use strict;

use Encode;
use LWP::UserAgent;
use XML::RSS;
use JSON;
use Data::Dumper;
use DateTime::Format::Mail;
use URI::Escape;
use HTML::Entities;
use FindBin;
use IPC::Run qw(run);

use utf8;

# See https://hn.algolia.com/api
my $jsonURL = 'https://hn.algolia.com/api/v1/search_by_date?tags=%28story,poll%29&numericFilters=points%3E100';

my $db = 'descriptions.db';

# my $db = 'foo.db';

my $ua = LWP::UserAgent->new(
   # ssl_opts => { verify_hostname => 0 },
);
$ua->agent("Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:61.0) Gecko/20100101 Firefox/61.0");

sub getFirehoseJSON {

    # Create a request
    my $req = HTTP::Request->new(GET => $jsonURL);

    # Pass request to the user agent and get a response back
    my $res = $ua->request($req);

    # Check the outcome of the response
    if (! $res->is_success) {
        die sprintf  "*Error* from GET: %s", $res->status_line;
    }
    return $res->content;
}

my $json = getFirehoseJSON();

# my $json;
# open(my $i, "ost2.json");
# {
#     local $/;
#     $json = <$i>;
# }
# close $i;

# print $res->content;
my $firehose = decode_json($json);

# my $d = Data::Dumper->new([$firehose]);
# $d->Sortkeys(1);
# $d->Indent(1);
# print $d->Dump();
# __END__

# my $rss = new XML::RSS (version => '1.0', encoding=>'ISO-8859-1');
my $rss = new XML::RSS (version => '1.0');
$rss->channel(
  title        => "Hacker News 100 - Readable Contents",
  description  => "by Peter Valdemar Mørch",
);
use DBI;
my $dbh = DBI->connect("dbi:SQLite2:dbname=$db","","");

my $getDescrSth = $dbh->prepare('
    SELECT description
      FROM descriptions
     WHERE objectID = ?
');

my $insertDescrSth = $dbh->prepare('
    INSERT INTO descriptions
        (objectID, description, createTime)
    VALUES (?,?,?)
');

sub getReadable {
    my ($url) = @_;
    printf STDERR "Getting %s\n", $url;
    my $exec = "$FindBin::Bin/python/bin/breadability";
    my $stdout;
    my $stderr;
    my $success = run(
        [ $exec, $url ],
        '>', \$stdout,
        '2>', \$stderr
    );
    my $readable;
    if ($success) {
        $readable = $stdout;
    } else {

        $readable = qq{Couldn't get $url with breadability.};
        if ($stdout) {
            $readable .= qq{ STDOUT: $stdout};
        }
        if ($stderr) {
            $readable .= qq{ STDERR: $stderr};
        }
    }
    return $readable;
}

sub getDescription {
    my ($hit) = @_;

    my $readable = getReadable(
        $hit->{url} ?  $hit->{url} : $hit->{hnewsUrl}
    );

    my $encURL = encode_entities($hit->{url});
    my $encHnewsURL = encode_entities($hit->{hnewsUrl});

    my $description = '<p>';
    if ($hit->{url}) {
        $description .= sprintf 'URL: <a href="%s">%s</a>, ', $encURL, $encURL;
    }
    $description .= sprintf qq(See on <a href="%s">Hacker News</a></p>\n%s\n),
        $encHnewsURL,
        $readable;
    return $description;
}

# binmode(STDOUT, ":utf8");
foreach my $hit (@{ $firehose->{hits} }) {
    my $hnewsUrl = sprintf "https://news.ycombinator.com/item?id=%d",
        $hit->{objectID};
    $hit->{hnewsUrl} = $hnewsUrl;

    $getDescrSth->execute($hit->{objectID});

    my ($description) = $getDescrSth->fetchrow_array();
    unless ($description) {
        $description = getDescription($hit);
        $insertDescrSth->execute($hit->{objectID}, $description, time);
    }
    my $descriptionChars = Encode::decode('UTF-8', $description);
    # $descriptionChars = 'æøå don’t in UTF8<br>' . $descriptionChars;
    my $dt = DateTime->from_epoch( epoch => $hit->{created_at_i} );
    my $dateStr = DateTime::Format::Mail->format_datetime( $dt );
    my %item = (
        title       => sprintf("%s%s (%d pts)",
                            $hit->{url} ? '' : 'HNInternal: ',
                            $hit->{title}, $hit->{points}),
        link        => $hit->{url} // $hnewsUrl,
        description => $descriptionChars,
        dc => {
            date => $dateStr
        }
    );
    # use Data::Dump qw(dump);
    # print dump(\%item), "\n";
    $rss->add_item(%item);
}

print $rss->as_string();

File: rss.js
#!/usr/bin/node
const { Readability } = require('@mozilla/readability');
const { JSDOM, VirtualConsole } = require('jsdom');
const axios = require('axios');
const process = require('process');
const fs = require('fs')
const { Feed } = require("feed");
const sqlite3 = require('sqlite3').verbose();
const util = require('util');
const he = require('he');

const jsonURL = 'https://hn.algolia.com/api/v1/search_by_date?tags=%28story,poll%29&numericFilters=points%3E100';
const detailLimit = 200;
const AXIOS_TIMEOUT = 10000

const dbFile = 'data/articles.db';
const db = new sqlite3.Database(dbFile)
db.serialize()

db.asyncRun = util.promisify(db.run);
db.asyncGet = util.promisify(db.get);
db.asyncAll = util.promisify(db.all);

const NULL_DEVICE='/dev/null';

async function initDatabase() {
    const existsRows = await db.asyncAll(`
        SELECT name
          FROM sqlite_master
         WHERE type='table'
           AND name='articles';
    `)
    if (existsRows.length == 0) {
        await db.asyncRun(`
            CREATE TABLE articles (
                objectID integer primary key,
                createTime integer,
                title text,
                points integer,
                description text,
                url text
            );
        `);
        await db.asyncRun(`
            CREATE INDEX createTimeIndex on articles(createTime);
        `);
    }
}

async function deleteOldArtcilesFromDatabase() {
    // 30 days ago
    await db.asyncRun(`
        DELETE FROM articles
        WHERE createTime < STRFTIME('%s') - 30*24*60*60;
    `);
}

function getJSDOM(text, url) {
    const virtualConsole = new VirtualConsole()
    for (let event of [ 'jsdomError', 'error', 'warn', 'info', 'dir' ]) {
        virtualConsole.on(event, (error) => {
            if (error.detail && error.detail.length &&
                error.detail.length > detailLimit) {
                error.detail =
                    `long string, length:${
                    error.detail.length}, first ${detailLimit
                    } chars:\n${error.detail.substring(0, detailLimit)}`
            }
            console.error("vconsole " + event, error)
        });
    }
    return new JSDOM(text, { virtualConsole, url })
}

async function readability(url) {
    let response
    try {
	response = await axios.get(url, {
            timeout: AXIOS_TIMEOUT
        })
    } catch(error) {
	return `Couldn't get ${url}: ${error}`
    }

    // See Raw PDF contents shown as readable contents · Issue #703 ·
    // mozilla/readability
    // https://github.com/mozilla/readability/issues/703
    if (! (response.headers['content-type'] &&
           response.headers['content-type'].match(/html|xml/)
    ))
        return '&lt;Not HTML&gt;'

    html = response.data
    var doc = getJSDOM(html, url)
    let reader = new Readability(doc.window.document);
    let article = reader.parse();
    doc.window.close();

    // Also see Raw PDF contents shown as readable contents · Issue #703 ·
    // mozilla/readability
    // https://github.com/mozilla/readability/issues/703
    if (article == null || (
        ! article.title &&
        ! article.byline &&
        ! article.dir &&
        ! article.excerpt &&
        ! article.siteName
    )) {
        return '&lt;Unparsable&gt;'
    }
    return article.content;
}

async function updateArticleInDB(objectID) {
    let row = await db.asyncGet(
        `SELECT points FROM articles WHERE objectID=?`,
        hit.objectID
    )
    if (! row)
        return false
    if (row.points != hit.points) {
        process.stderr.write(
            `Updating points ${row.points}->${hit.points} for: ${
                getArticleURL(hit)
            }\n`
        );
        db.asyncRun(
            `UPDATE articles
                SET points=?
              WHERE objectID=?`,
            hit.points,
            hit.objectID
        )
    }
    return true
}

function getHNewsURL(hit) {
    return `https://news.ycombinator.com/item?id=${hit.objectID}`
}

function getArticleURL(hit) {
    return hit.url ? hit.url : getHNewsURL(hit)
}

async function addHitToDB(hit) {
    const url = getArticleURL(hit)
    const hnewsURL = getHNewsURL(hit)

    process.stderr.write(`Getting: ${url}\n`);

    let description = '<p>';
    if (hit.url) {
        const encURL = he.encode(hit.url)
        description += `URL: <a href="${encURL}">${encURL}</a>, `
    }
    description +=
        `See on <a href="${he.encode(hnewsURL)}">Hacker News</a></p>\n`
    description += await readability(url);

    const title = `${hit.url ? '' : 'HNInternal: '}${hit.title}`

    await db.asyncRun(`
        INSERT INTO articles (
            objectID,
            createTime,
            title,
            points,
            description,
            url
        )
        VALUES(?,?,?,?,?,?)`,
        hit.objectID,
        hit.created_at_i,
        title,
        hit.points,
        description,
        url
    )
}

async function addArticleToFeed(feed, article) {
    const date = new Date(article.createTime * 1000)

    const title = `${article.title} (${article.points} pts)`
    // console.log("   ", date.toISOString())
    feed.addItem({
        title,
        id: article.objectID,
        link: article.url,
        description: article.description,
        date,
    })
}

async function start() {
    initDatabase()

    const response = await axios.get(jsonURL, {
        timeout: AXIOS_TIMEOUT
    });
    for (hit of response.data.hits) {
        // Don't add the same article twice
        if (await updateArticleInDB(hit)) {
            continue;
        }
        await addHitToDB(hit);
    }

    const feed = new Feed({
      title: "HN100 - Readable Contents",
      link: jsonURL,
      description: "Uses Readability to add bodies to the RSS feed",
      language: "en",
      author: {
        name: "Peter V. Mørch",
        email: "peter@morch.com",
        link: "https://www.morch.com"
      }
    });

    let articles = await db.asyncAll(
        `SELECT *
           FROM articles
          WHERE createTime > strftime('%s', 'now') - 24*60*60
       ORDER BY createTime DESC`
    )
    for (let article of articles) {
        addArticleToFeed(feed, article)
    }

    process.stdout.write(feed.rss2())

    deleteOldArtcilesFromDatabase()

    db.close()
}

async function processArgUrls(urls) {
    for (let url of urls) {
        console.log(url)
        console.log(await readability(url))
    }
}

let urls = [ ...process.argv ].slice(2);
if (urls.length > 0) {
    processArgUrls(urls)
} else {
    start()
}
